### PR TITLE
Compact storage for beamtransfer matrices

### PR DESCRIPTION
**Pull request by jrs65**
The advantage of this scheme is that it should be much more efficient for reading the BTMs back off disk.

### CODE REVIEW

File: drift/core/beamtransfer.py [hunk 1]
```diff
@@ -675,30 +515,29 @@ def _generate_mfiles(self, regen=False):
...
+
         # Calculate the number of baselines to deal with at any one time. Aim
         # to have a maximum of "mem_chunk" GB in memory at any one time
-        fbsize = (
-            self.telescope.num_pol_sky
-            * (self.telescope.lmax + 1)
-            * (2 * self.telescope.mmax + 1)
-            * 16.0
-        )
+        fbsize = self.telescope.num_pol_sky * nl * 2 * nm * 16.0
```
Comment (sjforeman) [inline drift/core/beamtransfer.py:540]: Should it be `np_inc` instead of `self.telescope.num_pol_sky` here?

**Review by sjforeman — APPROVED** [review]
Kudos for figuring out a nice scheme for this. I only have one comment about your changes.

Separately, I can see a few ways to expand on this scheme that might improve performance elsewhere in the pipeline:
- Updating `project_matrix_sky_to_telescope` in the same way as your new `project_vector_sky_to_telescope`
- Only doing the "telescope SVD" on the BTM in the "compact representation" (with skipped frequencies and baselines removed), and updating the corresponding projection routines. This could potentially give some nice savings in runtime and memory usage for these SVDs

These are only relevant for the KL pipeline, though, so we can leave them for later.

File: drift/core/beamtransfer.py (hunk 1 above)
Comment (jrs65) [reply drift/core/beamtransfer.py:540]: I *think* it actually still needs to be num_pol_sky for the moment anyway. The reason is that some of the internal steps still allocate the memory for the full num_pol_sky when generating the BTMs, and then it's only trimmed back to np_inc as it is about to write out to disk.

**Comment by jrs65** [comment]
Definitely agree that specifically utilising knowledge of the compact scheme could speed up the other projections. I just didn't want to mess them up here especially as I wasn't going to be doing much testing of them. I'll create an issue for that enhancement.

**Pull request merged**